File: computer-science/data-structures/abstract-data-type/queue/impl/queue.clj
(defprotocol IQueue
  (add-el [this element])
  (remove-el [this])
  (peek-el [this]))

(defrecord Queue [q]
  IQueue
  (add-el [_ el]
    (swap! q conj el))
  (remove-el [_]
    (swap! q next))
  (peek-el [_]
    (first @q)))

File: computer-science/data-structures/abstract-data-type/queue/impl/queue.js
class Queue {
  constructor() {
    this.data = [];
  }

  add(record) {
    this.data.unshift(record); // Add to the first place
  }

  remove() {
    return this.data.pop(); // Remove from the last place
  }

  get peek() {
    return this.data[this.data.length - 1];
  }
}
